validation: Improve error handling when VerifyDB dosn't finish successfully

mzumsande opened the pull request:
`VerifyDB()` can fail to complete due to insufficient dbcache at the level 3 checks. This PR improves the error handling in this case in the following ways:
- The rpc `-verifychain` now returns false if the check can't be completed due to insufficient cache
- During init, we only log a warning if the default values for `-checkblocks` and `-checklevel` are taken and the check doesn't complete. However, if the user actively specifies one of these args, we return with an InitError if we can't complete the check.

This PR also changes `-verifychain` RPC to return `false` if the verification didn't finish due to missing block data (pruning) or due to being interrupted by the node being shutdown.

Previously, this PR also included a fix for a possible assert during verification - this was done in #27009 (now merged).

saranshisatgit commented:
I have tested the PR the command `./bitcoin-cli -signet verifychain 4 1000` true , does it requires tests to be written ?

Raulnori77 commented:
https://prnt.sc/azRPNYbPq46J.diff

mzumsande commented:
Note that this doesn't only affect the `verifychain` RPC, but is also called as part of the startup (with the default level 3), which prevent the node from starting if validation fails.
I think because of that, the current strategy is to only return an error if we are sure something is invalid, but permit the application to start with just a warning in the debug log if we couldn't complete all checks. If only the RPC was affected, it would be ok to just return an error if we weren't able to do all the checks we asked for.

saranshisatgit commented:
> verifychain
The above methods aborts, the code does produce assertion and thus abort, are you suggesting that it should rather identify for the user to simply ignore or warn the user that there is an issue with the verification of the chain, since we are starting up?

mzumsande commented:
> are you suggesting that it should rather identify for the user to simply ignore or warn the user that there is an issue with the verification of the chain, since we are starting up?

This PR reduces the number of checks that are being made when there is not sufficient ccache memory, so that the assertion can't hit anymore (assertions can't be ignored in general).
What I was talking about is what `VerifyDB()` should do if it didn't completely verify the blocks at the levels required (but also didn't encounter an actual error), and I explained why I think it is ok to give a warning in the debug log.

saranshisatgit commented:
Right, so it should be taken care by default or by input, as in this there is an input , so that means a user is always triggering the output. 

Now, this is my suggestion, that we should perhaps make this part of the default startup and should throw some debut output and continue with the startup process.?

What do you think? @mzumsande

mzumsande commented:
> Now, this is my suggestion, that we should perhaps make this part of the default startup and should throw some debut output and continue with the startup process.?

Not sure I understand. `VerifyDB()` is already part of the default startup, and this PR adds a debug log entry if we skip some L3 checks because if our ccache is insufficient.

saranshisatgit commented:
I was under the assumption that you are talking about VerifyDB() handling the checks like the above on its own on the startup, rather the rpc doing the check !

fanquake commented on src/validation.cpp:
```diff
@@ -3975,30 +3976,36 @@ bool CVerifyDB::VerifyDB(
     }
-
+    if (skipped_l3_checks) {
+        LogPrintf("VerifyDB(): Skipped verification of level >=3 (insufficient database cache size). Consider increasing -dbcache.\n");
```
Maybe put `Skipped verification of...` on it's own line:
```bash
2022-10-03T14:00:20Z Verifying last 6 blocks at level 4
2022-10-03T14:00:20Z [0%]...[16%]...[25%]...[33%]...[41%]...[50%]...VerifyDB(): Skipped verification of level >=3 (insufficient database cache size). Consider increasing -dbcache.
2022-10-03T14:00:20Z [DONE].
```

mzumsande replied:
Will add a commit to do that - the downside is obviously is that it makes the progress logging more verbose (up to 10 lines instead of one).

mzumsande replied:
done

fanquake reviewed (approved):
ACK b7adbca2061e5a88130f61e2744885ab97340d5c - Could be improved in follow ups. i.e if a user has explictly set `-checklevel=3||4`, it's probably better to abort on too-small dbcache/where we can't actually perform the checks, rather than log a warning.

I also agree that changing `verifychain` to return warnings/errors is also a better behaviour than returning `true` (and just logging) when you've actually skipped level 3 & 4 checks.

mzumsande commented:
Happy to incorporate the suggestions from above here (no need for a follow-up).

mzumsande committed:
validation: Change return value of VerifyDB to enum type

This does not change behavior. It is in preparation for
special handling of the case where VerifyDB doesn't finish
for various reasons, but doesn't fail.

mzumsande committed:
init, validation: Improve handling if VerifyDB() fails due to insufficient dbcache

The rpc command verifychain now fails if the dbcache was not sufficient
to complete the verification at the specified level and depth.

In the same situation, the VerifyDB check during Init will now fail (and lead to
an early shutdown) if the user has explicitly specified -checkblocks or
-checklevel but the check couldn't be executed because of the limited
cache. If the user didn't change any of the two and is using the defaults, log a warning
but don't prevent the node from starting up.

mzumsande commented:
[57f367a ](https://github.com/bitcoin/bitcoin/commit/57f367a6f0ede978513beadfc084205a012e1a22)to [ad2a1f8](https://github.com/bitcoin/bitcoin/commit/ad2a1f82c8a134aa2dcf10e33551811c94113727):
I expanded this PR to incorporate the feedback (and rebased):
Additionally changed behavior:
- The rpc `-verifychain` now returns false if the check can't be completed due to insufficient cache.
- During init, we only log a warning if the default values for `-checkblocks` and `-checklevel` are taken and the check doesn't complete. However, if the user actively specifies one of these args, we return with an InitError.
- Logging of verification progress is split over multiple lines. This is more verbose, but now each update has its own timestamp, and other threads logging concurrently will no longer lead to mangled output.

Will adjust OP and title.

pablomartin4btc commented on src/rpc/blockchain.cpp:
```diff
@@ -1107,7 +1107,7 @@ static RPCHelpMan verifychain()
     Chainstate& active_chainstate = chainman.ActiveChainstate();
     return CVerifyDB().VerifyDB(
-        active_chainstate, chainman.GetParams().GetConsensus(), active_chainstate.CoinsTip(), check_level, check_depth);
+               active_chainstate, chainman.GetParams().GetConsensus(), active_chainstate.CoinsTip(), check_level, check_depth) == VerifyDBResult::SUCCESS;
```
nit: not sure about those extra spaces there; could you please double check what's mentioned in the [developer notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md) about [AlignAfterOpenBracket](https://clang.llvm.org/docs/ClangFormatStyleOptions.html) style option?

mzumsande replied:
I ran the Clang formatter script over the commits (as described in `/contrib/devtools/README.txt`) and I'm pretty sure it added the space for me.

pablomartin4btc reviewed (commented):
Tested ACK, I've managed to reproduce the issue before the fix as follows in the screenshots:

From the bitcoind execution, level 3 & 4 from cli (which crashes it)
![pr 25574 - bitcoind - before fix](https://user-images.githubusercontent.com/110166421/205998145-4c562ced-17bc-4ee0-bdc0-9eba7d930e80.JPG)

From the client terminal, level 3 or 4 both return true (this one before it crashes)

![image](https://user-images.githubusercontent.com/110166421/205999327-0fcf90db-36a4-445d-ac19-7680a5508583.png)


![pr 25574 - cli - before fix](https://user-images.githubusercontent.com/110166421/205998856-60015aba-90a6-4e92-a122-d3d8a7605c99.JPG)


After applying the fix, both levels return false:

![image](https://user-images.githubusercontent.com/110166421/205999574-bd7b4655-d18f-4a52-b92f-64d62acb3b76.png)


And bitcoind doesn't crash anymore (capturing both levels):

![image](https://user-images.githubusercontent.com/110166421/206000231-c9c9cfca-e00f-4ec7-83d3-fe30634db009.png)

LarryRuane commented:
Just FYI, I was unable to reproduce the problem with `bitcoin-cli -signet verifychain 4 1000`. However, it did reproduce when I increased the number of blocks to 5000. I'll review this PR soon.

hernanmarino reviewed (commented):
tested ACK , everything running as expected.

MarcoFalke commented on src/node/chainstate.cpp:
```diff
@@ -166,10 +166,11 @@ ChainstateLoadResult VerifyLoadedChainstate(ChainstateManager& chainman, const C
+                *chainstate, chainman.GetConsensus(), chainstate->CoinsDB(),
+                options.check_level,
+                options.check_blocks);
+            if (result == VerifyDBResult::ERROR_CORRUPTED_BLOCK_DB) {
```
e7dfc743406bfa0cac94c640dd5eaeac4364c990: Should be switch-case to avoid missing an error value?

mzumsande replied:
changed to use `switch`.

MarcoFalke commented on src/validation.h:
```diff
@@ -343,12 +343,17 @@ bool HasValidProofOfWork(const std::vector<CBlockHeader>& headers, const Consens
     CVerifyDB();
     ~CVerifyDB();
-    bool VerifyDB(
+    VerifyDBResult VerifyDB(
```
e7dfc743406bfa0cac94c640dd5eaeac4364c990: should be nodiscard?

mzumsande replied:
done

MarcoFalke commented on src/init.cpp:
```diff
@@ -1438,6 +1438,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         options.prune = node::fPruneMode;
         options.check_blocks = args.GetIntArg("-checkblocks", DEFAULT_CHECKBLOCKS);
         options.check_level = args.GetIntArg("-checklevel", DEFAULT_CHECKLEVEL);
+        options.fail_on_insufficient_dbcache = args.IsArgSet("-checkblocks") || args.IsArgSet("-checklevel");
```
42402bffda6643346cb2b51c9e85b389f3d19068: Any reason for this bool? The `DEFAULT_CHECKLEVEL` is `3`, presumably because `4` would almost never work anyway. And with `3` the check can't fail due to insufficient cache. Seems better to remove this bool and "replace" it with `true`?

MarcoFalke replied:
Ah, I see previously it would silently pass, but in this pull you changed it to an error

MarcoFalke replied:
I wonder if this requires a release note, or alternatively if the patch would be smaller and easier to review if it only fixed the level 4 issue and left level 3 as-is for now?

mzumsande replied:
I added a release note, for the rest see the top-level comment below.

MarcoFalke commented on src/node/chainstate.h:
```diff
@@ -35,7 +36,11 @@ struct ChainstateLoadOptions {
+                                  FAILURE,
+                                  FAILURE_INCOMPATIBLE_DB,
+                                  INTERRUPTED,
+                                  FAILURE_INSUFFICIENT_DBCACHE };
```
42402bffda6643346cb2b51c9e85b389f3d19068 nit: Maybe order this before/after `FAILURE_INCOMPATIBLE_DB` to group related things into one section?

mzumsande replied:
done

MarcoFalke commented on src/validation.h:
```diff
@@ -346,6 +346,7 @@ arith_uint256 CalculateHeadersWork(const std::vector<CBlockHeader>& headers);
 enum class VerifyDBResult {
     SUCCESS,
     ERROR_CORRUPTED_BLOCK_DB,
+    ERROR_INSUFFICIENT_CACHE
```
42402bffda6643346cb2b51c9e85b389f3d19068 nit : missing trailing `,`?

MarcoFalke commented on src/validation.cpp:
```diff
@@ -4028,30 +4029,36 @@ bool CVerifyDB::VerifyDB(
     }
-
+    if (skipped_l3_checks) {
+        LogPrintf("VerifyDB(): Skipped verification of level >=3 (insufficient database cache size). Consider increasing -dbcache.\n");
```
nit 46ec70affaf6ea9d33a9c403b605c0fb5a0c1d6c: No need for the function name in newly added log message, since `-logsourcelocations` exists.

mzumsande replied:
removed the function name.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -4010,19 +4010,22 @@ bool CVerifyDB::VerifyDB(
         if (nCheckLevel >= 1 && !CheckBlock(block, state, consensus_params)) {
-            return error("%s: *** found bad block at %d, hash=%s (%s)\n", __func__,
-                         pindex->nHeight, pindex->GetBlockHash().ToString(), state.ToString());
+            LogPrintf("%s: *** found bad block at %d, hash=%s (%s)\n", __func__,
```
nit in e7dfc743406bfa0cac94c640dd5eaeac4364c990: (here and elsewhere)

When changing this from `error`, which included `ERROR: ` in the log message, it could make sense to keep some kind of mention of "error" and instead remove the function name, which the user can already inject via `-logsourcelocations`.

maflcko reviewed (approved):
review ACK a1c83166f33d44a0909d8789d3a4bc8b2539842e 💁

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK a1c83166f33d44a0909d8789d3a4bc8b2539842e 💁
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh5hgv9Eu0s1Ko/vpJoWfVuakYVEU7HmFMAqGM+iJtFO+eWSciUcfyMdMY+Pi7L
rS76o5lIRufk5x6CE7PaeOlxpfbMbKy7AEj13ps4cJiXEtFw/QPW4IhfGxgbxp8e
z7GBk7R9UzRhDYz3DF3MNp5M4m3xvQ/fQb2VlQKtjSgQ65vD4c1oLLhp48bMj/+B
/QNVUOw0L6AsSI6EDjp3SWviDtfXyMxQiCpmkS1t+iM5rAuVXsJVfSUldTP5V6SP
QFa5yNoomUIYXVwAOs1dHwj2G8r15r/wzZH8ucLz2JbGYiTsGgJszUhmrF1SLJlD
cTGZnw4/CI9bbiRKQj+Cbp/bZiwwYqe972w7OBugWlqp8YM5xYAcrjVRFUHzn15B
CvW6+LGE1QMQZSPRdjkea0F48CtxJ43WD8udtDNtor4Ehh7Lwx5BjQ3lhIsWa3Gp
LejMpwMO2ceG8TQaibc80ARA/ahg2YQpqyoZW+KJDMzqgTvTKJaBn0fq07j1RzEu
UodOx8e6
=Jqoz
-----END PGP SIGNATURE-----
```


</details>

MarcoFalke commented on src/validation.cpp:
```diff
@@ -3986,15 +3986,15 @@ VerifyDBResult CVerifyDB::VerifyDB(
     BlockValidationState state;
```
nit in a1c83166f33d44a0909d8789d3a4bc8b2539842e: Since this commit is a fixup (https://github.com/bitcoin/bitcoin/pull/25574#discussion_r985851229) to the first commit, it can be re-ordered right after the first commit, if you re-touch.

mzumsande replied:
done

maflcko reviewed (commented):
left more comments (feel free to ignore)

mzumsande committed:
doc: add release note for #25574

mzumsande commented:
[a1c8316 ](https://github.com/bitcoin/bitcoin/commit/a1c83166f33d44a0909d8789d3a4bc8b2539842e)to [a96f7cb](https://github.com/bitcoin/bitcoin/commit/a96f7cb551b005aa5f6e68baaa8ba9806bcc504e): Addressed feedback by @MarcoFalke .

> I wonder if this requires a release note, or alternatively if the patch would be smaller and easier to review if it only fixed the level 4 issue and left level 3 as-is for now?

I still think that both parts make sense - fixing the assert fail, and changing the error handling to inform the user that the desired checks were not executed in the case of an insufficient cache instead of continuing silently.

But if reviewers would prefer it, I could split this into two parts: Reduce this PR to the first two commits (or maybe just the first), and then have a follow-up PR for the rest? Opinions?

fanquake commented:
I'm in favor of this change as-is. I'd rather we start returning errors / fail to startup, and let the user know that they are requesting something impossible given their system, than continue to fail silently/not perform the requested checks.

I think highlighting this in a release note is ok, and if some users have to modify their .conf files/runtime options, because their previously set arguments weren't actually acheiving anything (silent fail), that is also fine.

MarcoFalke commented:
> because their previously set arguments weren't actually acheiving anything (silent fail)

No, previously it was a silent *pass* on `verifychain` level 3. This pull changes it to the same type of fail (boolean value `false`) that would be occurring when the user ran into disk corruption. Not sure how user friendly that is. I am starting to think that if we want to return more error types from the rpc, it might be better to not cast every possible type into a boolean value.

fanquake commented:
> No, previously it was a silent pass on verifychain level 3.

Right, I misspoke, but I'd also consider it a "fail" for us to silently pass, but not actually do anything, if a user tries to perform a verification operation. I'm not sure why that's something we should continue to support, as opposed to telling the user it's not possible. In the worst case, at least it'd start exposing to users that calls they thought were previously doing something / meant something, actually don't.

> I am starting to think that if we want to return more error types from the rpc, it might be better to not cast every possible type into a boolean value.

Yea, a more explanatory error would be better as opposed to false.

MarcoFalke commented:
> Yea, a more explanatory error would be better as opposed to false.

That's a breaking RPC change, because a call that previously returned true now returns false (or an error message). If the RPC interface is broken, it appears more user-friendly to do only once. (Again, not asking the author to do here, but including the breaking commit here could cause breaking changes to be split over several releases.)

fanquake commented:
> That's a breaking RPC change, because a call that previously returned true now returns false (or an error message).

I think breaking the RPC would be fine, given the currently returned result is potentially meaningless/unreliable. I'm wondering how much usage `verifychain` would actually get in the wild.

mzumsande commented:
> That's a breaking RPC change, because a call that previously returned true now returns false (or an error message).

The RPC doc for `verifychain` says: 
```
Verifies blockchain database.
(...)
Result: 
true|false    (boolean) Verified or not
```
If the check didn't finish, the blockchain database was not verified, so the documentation is more in line with the behavior after this change than with the current behavior.

mzumsande commented:
To make some progress, I decided to split this into two: I will soon open another PR just for the assertion error fix (would be nice to at least have this fix in 25.0 imo) and put this PR into draft / reopen it after that other PR gets merged.

MarcoFalke commented:
Needs rebase

mzumsande committed:
validation: report if pruning prevents completion of verification

Now the verifychain RPC returns false if the checks didn't
finish because the blocks requested to be queried have been pruned.

mzumsande commented:
Rebased and put out of draft.

In the discussion of #27009, the issue of checks being left unfinished due to no more data being available in pruning mode came up:
> How is logging a warning for the not-enough-memory case different from logging a warning for the not-enough-storage case (pruning)? I think the warning can be displayed more prominently in a follow-up, but I think doing the same for similar cases seems fine. (https://github.com/bitcoin/bitcoin/pull/27009#issuecomment-1418721284)

So the question is should the behavior be analogous in the case where the checks are left uncompleted due to pruning? That is, should we both abort init in this case (if `-checkblocks` and `-checklevel` were overwritten by the user) and also return `false` in the RPC? (@john-moffett @MarcoFalke ).

I kind of think that there is a small difference between the two cases: If it's due to `-dbcache`, the query makes sense and the user can react by increasing their dbcache, with pruning it might make sense to handle this case gracefully, e.g. by not aborting init.

john-moffett commented:
> should we both abort init in this case (if `-checkblocks` and `-checklevel` were overwritten by the user) 

It's a close call for me. Given that the "prune failure" is unlikely to occur except on `-checkblocks` pretty substantially deep (at least 288, right?), I think it probably makes more sense to abort, since presumably they have a reason for it to be that deep. I suppose it also should be contingent on only having `-checkblocks` set. 

Again, though, it's a close call. I'd be okay with just a more detailed warning (eg - last `n` blocks requested to be checked, but only `m` are available due to pruning).

> and also return `false` in the RPC?

I feel pretty strongly that we should, assuming we keep the boolean result, which I'm okay with. I'd also be okay if it were changed to a more expressive result, but maybe it doesn't need the additional complexity.

mzumsande commented:
[f5724d1 ](https://github.com/bitcoin/bitcoin/commit/f5724d18878c292e4cdc766ac93d28d4252dd157)to [080f83e](https://github.com/bitcoin/bitcoin/commit/080f83e43ddc1fd2ddc57f6590b9166a373d38c2):

I added a commit to change the make the RPC return `false` in the case of unavailable data and improved the error messages. I left `init` behavior unchanged in this case (no aborts).

john-moffett commented:
ACK https://github.com/bitcoin/bitcoin/commit/080f83e43ddc1fd2ddc57f6590b9166a373d38c2 apart from the incidental test failure, which you've probably already caught:

```diff
diff --git a/test/functional/feature_pruning.py b/test/functional/feature_pruning.py
index 664ed779db2..03f5f2c8037 100755
--- a/test/functional/feature_pruning.py
+++ b/test/functional/feature_pruning.py
@@ -226 +226 @@ class PruneTest(BitcoinTestFramework):
-        with self.nodes[2].assert_debug_log(expected_msgs=['block verification stopping at height', '(pruning, no data)']):
+        with self.nodes[2].assert_debug_log(expected_msgs=['block verification stopping at height', '(no data)']):
```

mzumsande commented:
> which you've probably already caught

I hadn't, thanks for noticing! Fixed.

john-moffett reviewed (approved):
Re-ACK 42b192f0cbf9c04da111145c921344b0881b3ce3

MarcoFalke commented on doc/release-notes-25574.md:
```diff
@@ -0,0 +1,12 @@
+
+If the `-checkblocks` or `-checklevel` options are explicitly provided by the
+user, but the verification checks cannot be completed due to an insufficient
+dbcache, bitcoind will now return an error at startup. (#25574)
```
nit in 42b192f0cbf9c04da111145c921344b0881b3ce3: 
```suggestion
dbcache, Bitcoin Core will now return an error at startup. (#25574)
```

(holds for bitcoin-qt as well)

mzumsande replied:
done

maflcko reviewed (approved):
Left some non-blocking nits. Please let us know if you want to get this merged or address them.

review ACK 42b192f0cbf9c04da111145c921344b0881b3ce3 🗽

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 42b192f0cbf9c04da111145c921344b0881b3ce3 🗽
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUifdwwAvc/JCJeayQbrEGp/U6K39POo5P/O9usm+J6g4ljlFy6hfoqnEjhvDbo6
rEfXUf9ZlxR466QmiI2/6+2g9IUmPlrNB3vwD8MM3XJulYtsSfFtFarUwf6Z4Qip
eQjVmaC69aJJ0/B+ZGMsj5fLF6+uA3DPQxlS5+ceGFn2Dgqttmas33hurQMQbNIu
0hUD0NwZSeHeCrdzlAyllIJDSouMvjHkwmifIeyt/St0z5F6wo4z6QqkGWQkPDGK
Do+rUvDFSej2BeXW/LDKboHMBdph8YtPp4yvoDKnbthNfiRv+N2u4a7rQnbMtrLV
oFsNSc4H74lYsE9lCLRzrK/qDIDrXXWL5/f1GRfMSVSNQphNOomM28V4yBiPeCR7
T4v4G+WcsMI+lK9++dWk2J6J4d9eM5W09z/OJcGG271ARjdM38VppXxJzYQ9SPGw
3DqZfaGO8LEin5NOzT8i/NYOGdIyVCiu6WsD7PhFPa4PKimDXi5Ve+Ns1Hi1wYQ0
rBqhvH0A
=RQBu
-----END PGP SIGNATURE-----
```


</details>

mzumsande commented:
> Please let us know if you want to get this merged or address them.

I'll address the comments later this week!

ryanofsky commented on src/test/util/setup_common.cpp:
```diff
@@ -218,6 +218,7 @@ void TestingSetup::LoadVerifyActivateChainstate()
     options.prune = chainman.m_blockman.IsPruneMode();
     options.check_blocks = m_args.GetIntArg("-checkblocks", DEFAULT_CHECKBLOCKS);
     options.check_level = m_args.GetIntArg("-checklevel", DEFAULT_CHECKLEVEL);
+    options.fail_on_insufficient_dbcache = false;
```
In commit "init, validation: Improve handling if VerifyDB() fails due to insufficient dbcache" (cf307244db97e7b9e29a3d0ff948a6fa7877b548)

This block of code seems to duplicate a block of code in `AppInitMain`, but in the `AppInitMain` block this PR is setting `fail_on_insufficient_dbcache` to `args.IsArgSet("-checkblocks") || args.IsArgSet("-checklevel")`, while here it is being set to `false`. Probably both blocks of code should be deduplicated and call a shared function later. But for now it would be better to avoid creating unnecessary differences in the two blocks of code. Or if there is a reason test code is doing something different, there should be a code comment explaining why.

MarcoFalke replied:
Maybe just drop the hunk? Shouldn't matter either way :man_shrugging:

mzumsande replied:
changed to use same setting as in `AppInitMain`.

ryanofsky commented on src/validation.cpp:
```diff
@@ -4142,14 +4146,16 @@ bool CVerifyDB::VerifyDB(
             }
         }
-        if (ShutdownRequested()) return true;
+        if (ShutdownRequested()) return VerifyDBResult::SUCCESS;
```
In commit "validation: Change return value of VerifyDB to enum type" (422127b696f390e7cd34e897e7d74f09e3dae2cb)

In the interest of preventing future bugs where shutdown could be interpreted as success, it seems like it would be less misleading (and easy) to return a `VerifyDBResult::INTERRUPTED` error code here, analogous to `ChainstateLoadStatus::INTERRUPTED`.

mzumsande replied:
Good point! I added a commit to do this and adjusted the release notes (since this also changes the behavior of the `-verifydb` RPC)

ryanofsky reviewed (approved):
Code review ACK 42b192f0cbf9c04da111145c921344b0881b3ce3 with some suggestions and bikeshedding (feel free to ignore)

ryanofsky commented on src/node/chainstate.h:
```diff
@@ -25,6 +25,7 @@ struct ChainstateLoadOptions {
     bool reindex{false};
     bool reindex_chainstate{false};
     bool prune{false};
+    bool fail_on_insufficient_dbcache{false};
```
In commit "init, validation: Improve handling if VerifyDB() fails due to insufficient dbcache" (cf307244db97e7b9e29a3d0ff948a6fa7877b548)

`true` might be a safer default for libbitcoinkernel. Also the name of this option `ChainstateLoadOptions::fail_on_insufficient_dbcache` seems a little misleading because it seems like it is referring to insuffcient dbcache for loading, not insufficient_dbcache for verification. Would maybe change it to `ChainstateLoadOptions::require_full_verification` with a comment that setting the option to false can allow using a smaller dbcache size.

mzumsande replied:
changed to `ChainstateLoadOptions::fail_on_insufficient_dbcache`

I don't understand the comment. It's not an option the user can set, it's determined by `-checkblocks` and `-checklevel`. So if a user wants to run with an extremely low dbcache, they should change those parameters (or just use the default so that init will only warn).

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/25574#discussion_r1109110569

Thanks for implementing the suggestion. This comment was not about `bitcoind`, but about code that uses libbitcoinkernel such as:

https://github.com/bitcoin/bitcoin/blob/fe1b3256888bd0e70d0c9655f565e139ec87b606/src/bitcoin-chainstate.cpp#L93-L96

Code like this has it's own options and defaults, and is not aware of `checkblocks` or `checklevel`. It is just calling and using the `LoadChainstate` function directly. From the perspective of a `LoadChainstate` caller, I think `fail_on_insufficient_dbcache` could have been misleading because it sounds like it would allow loading with an insufficient dbcache, not just verifying with an insufficient dbcache. I do think in either case it wouldn't be bad for the option to have a code comment, so I suggested one below.

mzumsande committed:
validation: return VerifyDBResult::INTERRUPTED if verification was interrupted

This means that the -verifydb RPC will now return false if it
cannot finish due to the node being shutdown.

mzumsande commented on test/functional/feature_pruning.py:
```diff
@@ -223,7 +223,7 @@ def reorg_test(self):
         assert_raises_rpc_error(-1, "Block not available (pruned data)", self.nodes[2].getblock, self.forkhash)
-        with self.nodes[2].assert_debug_log(expected_msgs=['block verification stopping at height', '(pruning, no data)']):
+        with self.nodes[2].assert_debug_log(expected_msgs=['block verification stopping at height', '(no data)']):
             self.nodes[2].verifychain(checklevel=4, nblocks=0)
```
done

mzumsande commented on src/rpc/blockchain.cpp:
```diff
@@ -1110,7 +1110,7 @@ static RPCHelpMan verifychain()
                 },
                 RPCResult{
-                    RPCResult::Type::BOOL, "", "Verified or not"},
+                    RPCResult::Type::BOOL, "", "Verification finished successfully. If false, check debug.log for reason."},
```
That would break the existing API though - there might be users out there who'd need to adjust automated scripts if -`verifychain` returned a more complicated JSON object instead of a bool. 
I think I'd ACK a pull that does that, but would prefer not to include it here unless people really want it.

mzumsande commented on src/node/chainstate.h:
```diff
@@ -35,7 +36,12 @@ struct ChainstateLoadOptions {
+                                  FAILURE,
+                                  FAILURE_INCOMPATIBLE_DB,
+                                  FAILURE_INSUFFICIENT_DBCACHE,
+                                  INTERRUPTED,
```
fixed, though my local clang-format didn't complain before.

mzumsande commented on src/validation.cpp:
```diff
@@ -4154,13 +4154,14 @@ VerifyDBResult CVerifyDB::VerifyDB(
     }
     if (skipped_l3_checks) {
         LogPrintf("Skipped verification of level >=3 (insufficient database cache size). Consider increasing -dbcache.\n");
+        return VerifyDBResult::ERROR_INSUFFICIENT_CACHE;
```
Done - decided to use an if instead of ternary because `skipped_no_block_data` is treated similarly.
I decided to give `skipped_l3_checks` precedence - so if in addition to this, `skipped_no_block_data` is also true, we'd  report `VerifyDBResult::SKIPPED_L3_CHECKS`.

mzumsande commented on src/validation.h:
```diff
@@ -349,12 +349,19 @@ bool HasValidProofOfWork(const std::vector<CBlockHeader>& headers, const Consens
+    SUCCESS,
+    ERROR_CORRUPTED_BLOCK_DB,
+    ERROR_INSUFFICIENT_CACHE,
+    INSUFFICIENT_BLOCK_DATA,
```
done, using the latter suggestion.

mzumsande commented:
[42b192f ](https://github.com/bitcoin/bitcoin/commit/42b192f0cbf9c04da111145c921344b0881b3ce3)to [0af16e7](https://github.com/bitcoin/bitcoin/commit/0af16e7134459e0820ab95d751093876c1ec4c6d):

Addressed comments by @MarcoFalke and @ryanofsky, thanks for the reviews!
Also changed title and adjusted OP since this is no longer just about `-dbcache`-related error handling.

ryanofsky commented on src/node/chainstate.h:
```diff
@@ -25,6 +25,7 @@ struct ChainstateLoadOptions {
     bool reindex{false};
     bool reindex_chainstate{false};
     bool prune{false};
+    bool require_full_verification{true};
```
In commit "init, validation: Improve handling if VerifyDB() fails due to insufficient dbcache" (0c7785bb2540b69564104767d38342704230cbc2)

Might be worth adding a comment to explain why you would set the option. Maybe:

```c++
//! Setting require_full_verification to true will require all checks at
//! check_level (below) to succeed for loading to succeed. Setting it to
//! false will skip checks if cache is not big enough to run them, so may be
//! helpful for running with a small cache.
```

ryanofsky commented on src/validation.cpp:
```diff
@@ -4106,19 +4106,22 @@ bool CVerifyDB::VerifyDB(
         // check level 0: read from disk
         if (!ReadBlockFromDisk(block, pindex, consensus_params)) {
-            return error("VerifyDB(): *** ReadBlockFromDisk failed at %d, hash=%s", pindex->nHeight, pindex->GetBlockHash().ToString());
+            LogPrintf("Verification error: ReadBlockFromDisk failed at %d, hash=%s\n", pindex->nHeight, pindex->GetBlockHash().ToString());
```
In commit "validation: Change return value of VerifyDB to enum type" (6360b5302d2675788de5c4a28ea77d823f6d809e)

Commit message says this does not change behavior, but this is changing log prints slightly replacing `VerifyDB():` prefix with `Verification error:`. This is probably a good change, but could be clarified in the commit message.

ryanofsky commented on src/node/chainstate.cpp:
```diff
@@ -193,6 +193,7 @@ ChainstateLoadResult VerifyLoadedChainstate(ChainstateManager& chainman, const C
                 options.check_blocks);
             switch (result) {
             case VerifyDBResult::SUCCESS:
+            case VerifyDBResult::INTERRUPTED:
```
In commit "validation: return VerifyDBResult::INTERRUPTED if verification was interrupted" (d6f781f1cfcbc2c2ad5ee289a0642ed00386d013)

It would be less surprising to have `VerifyDBResult::INTERRUPTED` return `ChainstateLoadStatus::INTERRUPTED` than `ChainstateLoadStatus::SUCCESS`. Another potentially good side effect would be that it would avoid the caller printing a misleading load time if the load didn't really succeed:

https://github.com/bitcoin/bitcoin/blob/fe1b3256888bd0e70d0c9655f565e139ec87b606/src/init.cpp#L1520-L1524

mzumsande replied:
Makes sense - I'll open a small follow-up to address this in a few days.

ryanofsky reviewed (approved):
Code review ACK 0af16e7134459e0820ab95d751093876c1ec4c6d. Only small suggested changes since the last review, like renaming some of the enum values. I did leave more suggestions, but they are not very important and could be followups

john-moffett reviewed (approved):
ACK 0af16e7134459e0820ab95d751093876c1ec4c6d

maflcko reviewed (approved):
lgtm re-ACK 0af16e7134459e0820ab95d751093876c1ec4c6d 🎚

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

lgtm re-ACK 0af16e7134459e0820ab95d751093876c1ec4c6d 🎚
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgCEAwAl2HOYuqTfTHatbl1G6IC16Itmtb0chf3GhxESu4SIE7uNvMXIXVK2WG1
k5Sz9TeW2POSCwZnJtxdAzhNKwzN6Q79yJ0KL+6SOMxX7j4CaWun5b1kJf1xOlQx
gHBqBN1zLW65Wfh1L3qa4nzKiWwKdJJ2iUAUiatdJGLxElYIox98uTdq9Z3PuSSb
Xu1ivXy/EdXu+nQMeogF5V+g/yZAdSKGYSgS2ldG8D7k8cvY5BxEM/0b1UAjYOrE
gDKR7Ac7tG6sIzdfv+oY1vdjzRPbaATf2LoQiFgv0x0tPECKX+R19/g5UrU3DCZ0
1HSsj5rcuPJpy5YrL517v7K/XgUJlUhUsg6Xx7gn+jvu0qRrlhsFCSpv0XLzn+0J
aIKCXWLTV+2bLcMcktl/0+c/iFpRiFHaO8q1E4ivBttoGf3cK8Jv1CFup/+yyPaK
Ld8w3sZXEwuASxMEI03GbFKb1o4RMQojSkm9PaimfU3frbUFLTtUiHnw6X6SMjZ4
8KN4cKiS
=q5XI
-----END PGP SIGNATURE-----
```


</details>

achow101 commented:
ACK 0af16e7134459e0820ab95d751093876c1ec4c6d
